src/backend/kvm/thread.rs

```diff
@@ -0,0 +1,146 @@
+
+        // Allocate the new memory
+        let pages = Map::map(size * npgs)
+            .anywhere()
+            .anonymously()
+            .known::<perms::ReadWrite>(Kind::Private)
+            .map_err(|e| e.err.raw_os_error().unwrap_or(libc::ENOTSUP))?;
+
+        let mut keep = self.keep.write().unwrap();
+
+        // Map the memory into the VM
+        let vaddr = keep
+            .map(pages, addr)
+            .map_err(|e| e.raw_os_error().unwrap_or(libc::ENOTSUP))?
+            .as_virt()
+            .start;
+
+        Ok([vaddr.as_u64().into(), 0.into()])
+    }
+
+    pub fn meminfo(&self, block: &mut Block) -> Result<[Register<usize>; 2], i32> {
+        let keep = self.keep.read().unwrap();
+
+        // The maximum number of memory slots possible for a virtual machine
+        // minus the ones which were already used.
+        let mem_slots = keep.kvm_fd.get_nr_memslots() - keep.regions.len();
+
+        // FIXME:
+        // Obsolete, if [host side syscall verification and address translation](https://github.com/enarx/enarx/issues/957)
+        // is implemented.
+        let virt_start = Address::from(keep.sallyport_block_start.as_mut_ptr());
+
+        let mem_info: MemInfo = MemInfo {
+            virt_start,
+            mem_slots,
+        };
+
+        let c = block.cursor();
+        c.write(&mem_info).map_err(|_| libc::ENOBUFS)?;
+
+        Ok([0.into(), 0.into()])
+    }
+}
+
+impl super::super::Thread for Thread {
+    fn enter(&mut self) -> Result<Command> {
+        let vcpu_fd = self.vcpu_fd.as_mut().unwrap();
+        match vcpu_fd.run()? {
+            VcpuExit::IoOut(KVM_SYSCALL_TRIGGER_PORT, data) => {
+                debug_assert_eq!(data.len(), 2);
+                let block_nr = data[0] as usize + ((data[1] as usize) << 8);
+
+                let block_virt = self.keep.write().unwrap().sallyport_blocks[block_nr]
+                    .take()
+                    .unwrap();
+
+                // If some other thread tried to use the same block, the above unwrap would have panicked.
+                let block = unsafe { &mut *block_virt.as_mut_ptr::<Block>() };
+
+                // To avoid clashing of rep and req in the union, clone the request
+                let req = unsafe { block.msg.req };
+
+                let ret = match i64::from(req.num) {
+                    SYS_ENARX_BALLOON_MEMORY => {
+                        let rep = self.balloon(&req).into();
+                        block.msg.rep = rep;
+                        Ok(Command::Continue)
+                    }
+
+                    SYS_ENARX_MEM_INFO => {
+                        block.msg.rep = self.meminfo(block).into();
+                        Ok(Command::Continue)
+                    }
+
+                    _ => Ok(Command::SysCall(block)),
+                };
+
+                self.keep.write().unwrap().sallyport_blocks[block_nr].replace(block_virt);
+                ret
+            }
+            #[cfg(debug_assertions)]
```

Comment:
Add empty line after the close brace, please.

Reply:
ok